plugins/amongus/amongus.cpp

```diff
@@ -5,122 +5,132 @@
-	const std::string name = "Among Us.exe";
-	const auto id          = Process::find(name, pids);
-	if (!id) {
-		return false;
-	}
-
-	game = std::make_unique< Game >(id, name);
-	if (!game->isOk()) {
-		game.reset();
-		return false;
-	}
+uint32_t mumble_getFeatures() {
+	return MUMBLE_FEATURE_POSITIONAL;
+}
 
-	const auto fields = game->clientFields();
-	if (!inGame(fields)) {
-		game.reset();
-		return false;
+uint8_t mumble_initPositionalData(const char *const *programNames, const uint64_t *programPIDs, size_t programCount) {
+	for (size_t i = 0; i < programCount; ++i) {
+		if (strcmp(programNames[i], "Among Us.exe") != 0) {
+			continue;
+		}
+
+		game = std::make_unique< Game >(programPIDs[i], programNames[i]);
+
+		const auto error = game->init();
+		if (error == Game::NONE) {
+			const auto fields = game->clientFields();
+			if (!game->isMultiplayer(fields)) {
+				game.reset();
+				return MUMBLE_PDEC_ERROR_TEMP;
+			}
+		} else {
+			game.reset();
+			return error == Game::TEMP ? MUMBLE_PDEC_ERROR_TEMP : MUMBLE_PDEC_ERROR_PERM;
+		}
+
+		return MUMBLE_PDEC_OK;
 	}
 
-	return true;
+	return MUMBLE_PDEC_ERROR_TEMP;
 }
 
-static const std::wstring longDesc() {
-	return std::wstring(L"Supports Among Us with context and identity support.");
+void mumble_shutdownPositionalData() {
+	game.reset();
 }
 
-static std::wstring description(L"Among Us");
-static std::wstring shortName(L"Among Us");
+bool mumble_fetchPositionalData(float *avatarPos, float *avatarDir, float *avatarAxis, float *cameraPos,
+								float *cameraDir, float *cameraAxis, const char **contextPtr,
+								const char **identityPtr) {
+	const auto fields = game->clientFields();
 
-static int tryLock1() {
-	return tryLock(std::multimap< std::wstring, unsigned long long int >());
-}
+	if (!game->isMultiplayer(fields)) {
+		return false;
+	}
 
-static void nullUnlock() {
-}
+	if (fields.gameState == GameState::Ended) {
+		return true;
+	}
 
-static MumblePlugin amongusPlug = { MUMBLE_PLUGIN_MAGIC, description, shortName, nullptr, nullptr, tryLock1,
-									nullUnlock,          longDesc,    fetch };
+	const auto playerControlFields = game->playerControlFields();
 
-static MumblePlugin2 amongusPlug2 = { MUMBLE_PLUGIN_MAGIC_2, MUMBLE_PLUGIN_VERSION, tryLock };
+	// Mumble | Game
+	// X      | X
+	// Y      | -
+	// Z      | Y
+	const auto position = game->playerPosition(playerControlFields);
+	avatarPos[0] = cameraPos[0] = position.x;
+	avatarPos[2] = cameraPos[2] = position.y;
 
-extern "C" MUMBLE_PLUGIN_EXPORT MumblePlugin *getMumblePlugin() {
-	return &amongusPlug;
-}
+	for (uint8_t i = 0; i < 3; ++i) {
+		avatarDir[i] = cameraDir[i] = 0.f;
+		avatarAxis[i] = cameraAxis[i] = 0.f;
+	}
 
-extern "C" MUMBLE_PLUGIN_EXPORT MumblePlugin2 *getMumblePlugin2() {
-	return &amongusPlug2;
+	*contextPtr = game->context(fields).c_str();
+
+	const auto playerFields = game->playerInfoFields(playerControlFields);
+	*identityPtr            = game->identity(fields, playerFields).c_str();
```

Comment:
I assume the `context` and `identity` functions store the returned string somewhere and return it by reference?

Reply:
Yes, in the `Game` class.